### ISSUE TITLE
Git hooks and database migrations

### ISSUE THREAD

**Issue by Isk1n** [Mixd/wp-deploy#90]
Why *"All .sql files are – and should stay – git ignored"* ? What are the reasons behind this? Why not to use pre-commit and post-merge / post-rewrite hooks to export / import db on local machine, thus keeping database in sync with code? And modify deploy scripts to propagate db state between environments?

p.s. I haven't try to implement this method yet.

**Comment by lukewhitehouse** [comment]
Hi @Isk1n thanks for your feedback. That is potentially an option, however, I think there's a good chunk of times where you don't need to deploy the database, just to code (and vice-versa).

**Comment by joelataylor** [comment]
I agree with @lukewhitehouse - I certainly wouldn't want a db sync with every commit!

**Comment by lukewhitehouse** [comment]
@joelataylor Yep, I think this is a very niche feature that doesn't really belong in the core IMO. Thanks for the suggestion thoug @Isk1n, feel free to fork and modify :)

**Comment by Isk1n** [comment]
Thank you for your responses, @lukewhitehouse, @joelataylor. I am considering to make a fork, but first I would like to clarify some questions.

*tl;dr: How to merge dbs? What workflow do you use? Do you know a way to track every db modification in wp? What about uploads, why not to store them on git too, if each uploads manipulation would automatically produce git commit?*

Here is my situation:
* I am new to wp.
* I develop alone a small site.
* There are a bunch of 3rd-party plugins, which store their stuff in db alongside with pages and posts content. So I do a lot of db modifications on local machine. Whereas clients database constantly grows on production server. I need a way to **merge local and production dbs**. db:push and db:pull seems do not help here (maybe I miss something).
* What else I don't quite understand is **workflow**. [wiki](https://github.com/Mixd/wp-deploy/wiki/Notes-on-WordPress-development) contains very little info: wp is in submodule, plugins and themes are through local machine, uploads are separately. What I am interesting in is a more detailed description of development process, something about branching model and versioning. Is it something like [this](https://github.com/brianstorti/git-hooks)?

OK. So the ultimate goal of my possible fork is to reduce user interactions to minimum. For now my flow looks like this:
```bash
git push
staging deploy
staging db:push
staging uploads:push
git checkout master
git merge development
git push
production deploy
production db:push
production uploads:push
```
(where `staging` and `production` are aliases to `bundle exec cap staging` and `bundle exec cap production` respectively
)

What I want is:
```bash
staging deploy
production deploy
```

And here is my thoughts how to achieve this:
* First of all I decided that git hooks are not very handy for db migrations. Instead I am thinking of using wp hooks to export db and autocommit after every db modification. I see two problems:
	1. I have never seen a plugin which **exports db at every modification**, what makes me think that there is a some catch.
	2. If I am going to stick to some git flow, where should the autocommit go?
* In order to reduce space requirement and to increase merge speed and readability (even writability), db tables should be exported to separate files.
* Environment specific strings (basically just server_url) should be replaced with stubs on export and back on import.
* **Uploads** probably also should be stored in git cache in a similar fashion.

I will gladly accept any suggestions. Also sorry for the offtopic, if you like I will open separate issue about workflow and documentation.

**Comment by clov3rly** [comment]
@Isk1n : Merging databases is a hard problem. I haven't found a good solution yet. Best I can come up with is to never update the production database manually... ALWAYS make edits in dev and push it live. But that only works if you're the only dev. Definitely not a solved problem.

To reduce the commands you type in the terminal, I actually just run a bash function. You can write your own, but this one works for me. The `$fg[color]` variables are color output codes, from [oh-my-zsh](https://github.com/robbyrussell/oh-my-zsh). I'm also running a simple timer over the deploy process.

You'd want to add a few commands at the end to run db:push and uploads:push, but definitely keep in mind that running a destructive database overwrite automatically can come back to bite you! On the other hand, it's a good way to enforce the rule that any db updates MUST be made automatically.

```bash
function push {
  # check that we have a Capfile
  if [ ! -f "./Capfile" ]; then
          echo "$fg[red]Try running this from the root of your project. \
(The same directory that contains the Capfile.)$reset_color"
  fi
  # grab start time
  start=$SECONDS
  echo "Pushing site to ${1-staging} server with capistrano (via git remote)."
  # Push to git repo
  echo "$fg[green]• Pushing site to git remote.$reset_color"
  git push
  # Run capistrano
  echo "$fg[green]• Running capistrano on ${1-internal} server.$reset_color"
  bundle exec cap ${1:-internal} deploy
  bundle exec cap ${1:-internal} db:push
  bundle exec cap ${1:-internal} uploads:push
  # calculate duration
  duration=$(( SECONDS - start))
  echo "Finished deployment in $duration seconds."
}
```

And you can call it as `push` (which defaults to `push staging`) or `push production`, etc.
